lib/phoenix_live_view/async.ex

```diff
@@ -0,0 +1,209 @@
+            raise ArgumentError, """
+            expected assign_async to return map of assigns for all keys
+            in #{inspect(keys)}, but got: #{inspect(assigns)}
+            """
+          else
+            {:ok, assigns}
+          end
+
+        {:error, reason} ->
+          {:error, reason}
+
+        other ->
+          raise ArgumentError, """
+          expected assign_async to return {:ok, map} of
+          assigns for #{inspect(keys)} or {:error, reason}, got: #{inspect(other)}
+          """
+      end
+    end
+
+    new_assigns =
+      Enum.flat_map(keys, fn key ->
+        case socket.assigns do
+          %{^key => %AsyncResult{ok?: true} = _existing} -> []
+          %{} -> [{key, AsyncResult.new(key, keys)}]
+        end
+      end)
+
+    socket
+    |> Phoenix.Component.assign(new_assigns)
+    |> run_async_task(keys, wrapped_func, :assign)
+  end
+
+  defp run_async_task(%Socket{} = socket, keys, func, kind) do
+    if Phoenix.LiveView.connected?(socket) do
+      socket = cancel_existing(socket, keys)
+      lv_pid = self()
+      cid = cid(socket)
+      ref = make_ref()
+      {:ok, pid} = Task.start_link(fn -> do_async(lv_pid, cid, ref, keys, func, kind) end)
+      update_private_async(socket, &Map.put(&1, keys, {ref, pid, kind}))
+    else
+      socket
+    end
+  end
+
+  defp do_async(lv_pid, cid, ref, keys, func, async_kind) do
+    try do
+      result = func.()
+      Channel.report_async_result(lv_pid, async_kind, ref, cid, keys, {:ok, result})
+    catch
+      catch_kind, reason ->
+        Process.unlink(lv_pid)
+        caught_result = {:catch, catch_kind, reason, __STACKTRACE__}
+        Channel.report_async_result(lv_pid, async_kind, ref, cid, keys, caught_result)
+        :erlang.raise(catch_kind, reason, __STACKTRACE__)
+    end
+  end
+
+  def cancel_async(%Socket{} = socket, %AsyncResult{} = result, reason) do
+    new_assigns = for key <- result.keys, do: {key, AsyncResult.error(result, reason)}
+
+    socket
+    |> Phoenix.Component.assign(new_assigns)
+    |> cancel_async(result.keys, reason)
+  end
+
+  def cancel_async(%Socket{} = socket, keys, _reason) do
+    case get_private_async(socket, keys) do
+      {_ref, pid, _kind} when is_pid(pid) ->
+        Process.unlink(pid)
+        Process.exit(pid, :kill)
+        update_private_async(socket, &Map.delete(&1, keys))
+
+      nil ->
+        raise ArgumentError, "unknown async assign #{inspect(keys)}"
+    end
+  end
+
+  def handle_async(socket, maybe_component, kind, keys, ref, result) do
+    case prune_current_async(socket, keys, ref) do
+      {:ok, pruned_socket} ->
+        handle_kind(pruned_socket, maybe_component, kind, keys, result)
+
+      :error ->
+        socket
+    end
+  end
+
+  def handle_trap_exit(socket, maybe_component, kind, keys, ref, reason) do
+    {:current_stacktrace, stack} = Process.info(self(), :current_stacktrace)
+    trapped_result = {:catch, :exit, reason, stack}
```

Comment:
The stacktrace here is from the LV, not the task. Therefore, another reason to normalize the :catch in the task is that in here you could simply write `{:exit, reason}` (since that's the normalized one).